subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/SkipEmptySourceFilesTaskExecuter.java

```diff
@@ -61,7 +64,7 @@ public void execute(TaskInternal task, TaskStateInternal state, TaskExecutionCon
                 boolean deletedFiles = false;
                 boolean debugEnabled = LOGGER.isDebugEnabled();
                 for (File file : outputFileSet) {
-                    if (file.isFile()) {
+                    if (buildOutputCleanupRegistry.isOutputOwnedByBuild(file) && file.isFile()) {
```

Comment:
Not sure why this didn't end up in the PR https://github.com/gradle/gradle/commit/30ea2d504a3d0af60a464c8d4453a227230217eb#commitcomment-23292902:

@wolfs:
> We have only been removing files all along. If we want to remove directories as well we need to take care not to remove overlapping output directories. This is why I added this test since I expected the behaviour to be broken (i.e. we would delete the whole output directory even if other tasks had outputs in there, too): 30ea2d5#diff-f93c8d43f4f19ba60ff9ab9227d91180R117

I think we should clean up directories as well, to make the cleanup behavior consistent. I'm fine with doing it in a separate PR if that makes it easier.

> Regarding file.isFile() being cheaper: I actually put this in the other way round since I expected isOutputOwnedByBuild to be faster, since it does not need to hit the disk, while File.isFile() needs to look at the file on disk. Am I mistaken?

Hm. `isOutputOwnedByBuild()` needs to resolve files first, whereas `File.isFile()` is bound to come from some OS-level cache, so I think it should be faster in general.